src/CypherTools.Web/Client/Pages/Cypher.razor

```diff
@@ -0,0 +1,165 @@
+                <th></th>
+                <th>Form</th>
+                <th>Form Description</th>
+            </tr>
+
+            @foreach (var form in cypher.Forms)
+            {
+                <tr>
+                    <td>
+                        <button @onclick="() => RemoveForm(form)">X</button>
+                    </td>
+                    <td>
+                        <input @bind="@form.Form" />
+                    </td>
+                    <td>
+                        <BFUTextField @bind-Value=@form.FormDescription @bind-Value:event="OnChange" AutoAdjustHeight="true" Multiline="true"></BFUTextField>
+                    </td>
+                </tr>
+            }
+        </table>
+
+    </fieldset>
+
+
+    <div>
+        <BFUTextField Label="Effect" @bind-Value=@cypher.Effect @bind-Value:event="OnChange" AutoAdjustHeight="true" Multiline="true"></BFUTextField>
+    </div>
+
+    <fieldset>
+        <legend>Effect Options</legend>
+        <button @onclick="AddOption">Add Effect Option</button>
+        <table class="table table-responsive-lg">
+            <tr>
+                <th></th>
+                <th>Start Range</th>
+                <th>End Range</th>
+                <th>Effect Description</th>
+            </tr>
+
+            @foreach (var effectOption in cypher.EffectOptions)
+            {
+                <tr>
+                    <td>
+                        <button @onclick="() => RemoveOption(effectOption)">X</button>
+                    </td>
+                    <td>
+                        <input @bind="effectOption.StartRange" />
+                    </td>
+                    <td>
+                        <input @bind="effectOption.EndRange" />
+                    </td>
+                    <td>
+                        @*<input @bind="@form.FormDescription" />*@
+                        <BFUTextField @bind-Value="@effectOption.EffectDescription" @bind-Value:event="OnChange" AutoAdjustHeight="true" Multiline="true"></BFUTextField>
+                    </td>
+                </tr>
+            }
+        </table>
+    </fieldset>
+
+    <div>
+        <label>
+            Source:
+            <input @bind="cypher.Source" />
+        </label>
+    </div>
+}
+
+
+@code {
+    [Parameter]
+    public int id { get; set; } = 0;
+
+    private CypherTools.Core.Models.Cypher cypher;
+
+    protected override async Task OnInitializedAsync()
+    {
+        if (id == 0)
+        {
+            cypher = new Core.Models.Cypher();
+            cypher.Name = "New Character";
+        }
+        else
+        {
+            Console.WriteLine("Got here");
+            try
+            {
+                cypher = await Http.GetFromJsonAsync<CypherTools.Core.Models.Cypher>($"api/Cyphers/{id}");
+            }
+            catch (AccessTokenNotAvailableException exception)
+            {
+                exception.Redirect();
+            }
+
+        }
+    }
+
+    private async Task SaveCypher()
```

Comment:
On a more serious note, why is this here instead of in the controller? Is that a razor thing?

Reply:
The controller code has the actual database work.  This is just the pages code to get and save the objects.